src/codechandler/HEVCCodecHandler.cpp

```diff
@@ -0,0 +1,79 @@
+#include "HEVCCodecHandler.h"
+
+HEVCCodecHandler::HEVCCodecHandler(AP4_SampleDescription* sd) : CodecHandler(sd)
+{
+  if (AP4_HevcSampleDescription* hevc =
+          AP4_DYNAMIC_CAST(AP4_HevcSampleDescription, m_sampleDescription))
+  {
+    m_extraData.SetData(hevc->GetRawBytes().GetData(), hevc->GetRawBytes().GetDataSize());
+    m_naluLengthSize = hevc->GetNaluLengthSize();
+  }
+}
+
+bool HEVCCodecHandler::ExtraDataToAnnexB()
+{
+  if (AP4_HevcSampleDescription* hevc =
+          AP4_DYNAMIC_CAST(AP4_HevcSampleDescription, m_sampleDescription))
+  {
+    const AP4_Array<AP4_HvccAtom::Sequence>& sequences = hevc->GetSequences();
+
+    if (!sequences.ItemCount())
+    {
+      kodi::Log(ADDON_LOG_WARNING, "No available sequences for HEVC codec extra data");
+      return false;
+    }
+
+    //calculate the size for annexb
+    size_t sz(0);
+    for (const AP4_HvccAtom::Sequence *b(&sequences[0]), *e(&sequences[sequences.ItemCount()]);
+         b != e; ++b)
+      for (const AP4_DataBuffer *bn(&b->m_Nalus[0]), *en(&b->m_Nalus[b->m_Nalus.ItemCount()]);
+           bn != en; ++bn)
+        sz += (4 + bn->GetDataSize());
+
+    m_extraData.SetDataSize(sz);
+    uint8_t* cursor(m_extraData.UseData());
+
+    for (const AP4_HvccAtom::Sequence *b(&sequences[0]), *e(&sequences[sequences.ItemCount()]);
+         b != e; ++b)
+      for (const AP4_DataBuffer *bn(&b->m_Nalus[0]), *en(&b->m_Nalus[b->m_Nalus.ItemCount()]);
+           bn != en; ++bn)
+      {
+        cursor[0] = cursor[1] = cursor[2] = 0;
+        cursor[3] = 1;
+        memcpy(cursor + 4, bn->GetData(), bn->GetDataSize());
+        cursor += bn->GetDataSize() + 4;
+      }
+    kodi::Log(ADDON_LOG_DEBUG, "Converted %lu bytes HEVC codec extradata",
+              m_extraData.GetDataSize());
+    return true;
+  }
+  kodi::Log(ADDON_LOG_WARNING, "No HevcSampleDescription - annexb extradata not available");
+  return false;
+}
+
+bool HEVCCodecHandler::GetInformation(kodi::addon::InputstreamInfo& info)
+{
+  if (!info.GetFpsRate())
+  {
+    if (AP4_HevcSampleDescription* hevc =
+            AP4_DYNAMIC_CAST(AP4_HevcSampleDescription, m_sampleDescription))
+    {
+      bool ret = false;
+      if (hevc->GetAverageFrameRate())
+      {
+        info.SetFpsRate(hevc->GetAverageFrameRate());
+        info.SetFpsScale(256);
```

Comment:
you know why value 256? (set also below)
i think could be a candidate for a constexpr in a unnamed namespace

Reply:
I can only guess that HEVC doesn't have a rate and scale, just a rate.
@kszaq would you be able to help us here?

Reply:
Unfortunately I don't know why 256. The constant was introduced in this commit: https://github.com/xbmc/inputstream.adaptive/commit/8bfd456a556ab43976c2023ffbd2eb2e48a965cf

It looks like HVCC atom does not provide fps scale, instead it gives only average or constant frame rate. The constant 256 scale value is probably defined in MPEG-4 Part 15.

Reply:
Ah ok, thanks!